### ISSUE TITLE
Speed up CRF (Wapiti) training on Apple Silicon M4

### ISSUE THREAD

**Issue by cboulanger** [cboulanger/grobid#1]
## Background

CRF model training with Wapiti takes many hours on Apple Silicon M4. This issue documents the root causes and concrete improvement options.

## Current situation

- `nbThreads: 0` is already set in `grobid-home/config/grobid.yaml` → Wapiti uses all 10 logical cores (4E + 6P on M4). ✓
- `libwapiti.dylib` for `mac_arm-64` is a **generic arm64 build** — no SIMD/NEON intrinsics, no OpenMP, no Apple Accelerate framework. Only pthreads.
- No Metal GPU path exists for CRF/Wapiti. GPU acceleration is only for DeLFT deep-learning models.

## Root cause: epsilon convergence threshold

The dominant cost driver is the epsilon convergence threshold. Most models in `grobid.yaml` are set to `0.0000001` (1e-7):

```yaml
epsilon: 0.0000001   # segmentation, header, citation, name-*, etc.
nbMaxIterations: 2000
```

This tells Wapiti's L-BFGS optimizer to iterate until the loss changes by less than **one ten-millionth** per step. Note that `fulltext` already uses `0.0001` (1000× looser) with no reported quality regression. The relationship between epsilon and wall-clock time is roughly logarithmic: going from 1e-7 to 1e-5 can cut training time by **5–10×**.

### 1. Relax epsilon (biggest win, config-only change)

| Use case | Suggested epsilon | Expected speedup |
|---|---|---|
| Production training | `0.00001` (1e-5) | 3–7× |
| Development / iteration | `0.0001` (1e-4) | 10–20× |
| Quick smoke test | `0.001` | 30–50× |

The tighter models (`segmentation`, `header`, `citation`, `name-*`, etc.) could be changed to `0.00001` as a starting point and validated against evaluation scores.

### 2. Recompile `libwapiti.dylib` with M4-specific flags (~10–30% additional)

The pre-built `libwapiti.dylib` misses Apple M4 SIMD tuning. L-BFGS inner loops (dot products, scaled vector adds) map well to NEON. Rebuild from source inside `nix develop`:

```bash
git clone --depth 1 https://github.com/kermitt2/Wapiti /tmp/wapiti
cd /tmp/wapiti
CFLAGS="-O3 -mcpu=apple-m4 -ffast-math -funroll-loops" make lib
cp build/libwapiti.dylib grobid-home/lib/mac_arm-64/libwapiti.dylib
```

For a build targeting all M-series (not just M4): `-mcpu=apple-m1` or `-march=armv8.5-a+simd`.

### 3. Reduce `nbMaxIterations` for development

Pair with a looser epsilon for fast iteration during training data development:

```yaml
epsilon: 0.0001
nbMaxIterations: 500   # instead of 1500–2000
```

### 4. Use incremental training during data development

When adding new training files, use `incremental: true` (the `-i` JAR flag) to start from the existing model rather than training from scratch. This dramatically speeds up the annotation/train/check loop, even if the final production model should always be trained from scratch.

### 5. (Advanced) Integrate Apple Accelerate/BLAS into Wapiti

L-BFGS involves repeated BLAS-level operations (dot products, DAXPY). Calling `cblas_ddot` / `cblas_daxpy` from Accelerate could yield an additional 2–3× speedup on Apple Silicon due to AMX co-processor support, but requires modifying Wapiti's C source.

## What will NOT help

- **Metal GPU** — no GPU path exists for Wapiti CRF. Metal is DeLFT-only.
- **More threads** — `nbThreads: 0` already uses all available cores.

**Comment by cboulanger** [comment]
## Correction: `make lib` does not work — use CMake instead

The build command in the \"Recompile libwapiti.dylib\" section is incorrect. The upstream `kermitt2/Wapiti` fork uses **CMake**, not a `make lib` target. Running `make lib` will fail with `No rule to make target 'lib'`.

The correct build steps:

```bash
git clone --depth 1 https://github.com/kermitt2/Wapiti /tmp/wapiti
cd /tmp/wapiti
mkdir build && cd build
cmake .. \
  -DCMAKE_C_FLAGS="-O3 -mcpu=apple-m4 -ffast-math -funroll-loops" \
  -DCMAKE_CXX_FLAGS="-O3 -mcpu=apple-m4 -ffast-math -funroll-loops" \
  -DCMAKE_BUILD_TYPE=Release
make wapiti
cp libwapiti.dylib /path/to/grobid/grobid-home/lib/mac_arm-64/libwapiti.dylib
```

**Note:** CMake requires JNI (the SWIG wrapper is a JNI library), so `JAVA_HOME` must point to a JDK before running `cmake`. Inside `nix develop` this should already be set.

For a build targeting all M-series chips instead of just M4, replace `-mcpu=apple-m4` with `-mcpu=apple-m1` or `-march=armv8.5-a+simd`.
